back/sssscs/src/main/java/com/ib/certificate/CertificateController.java

```diff
@@ -0,0 +1,128 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.ib.certificate.CertificateRequest.Status;
+import com.ib.certificate.dto.CertificateRequestCreateDto;
+import com.ib.certificate.dto.CertificateRequestDto;
+import com.ib.certificate.dto.CertificateSummaryItemDto;
+import com.ib.user.IUserService;
+import com.ib.user.User;
+
+@RestController
+@RequestMapping("/api/cert")
+public class CertificateController {
+	@Autowired
+	private ICertificateService certificateService;
+	
+	@Autowired
+	private ICertificateRequestService certificateRequestService;
+	
+	@Autowired
+	private IUserService userService;
+	
+	@PostMapping("/request")
+	public ResponseEntity<String> makeRequest(/*@DTO(CertificateRequestCreateDto.class)*/ @RequestBody CertificateRequestCreateDto certificate) {
+		// TODO: Automatic mapping.
+		
+		CertificateRequest req = new CertificateRequest();
+		req.setIssuer(userService.findById(certificate.getIssuerId()));
+		req.setType(certificate.getType());
+		req.setValidTo(certificate.getValidTo());
+		req.setStatus(Status.PENDING);
+		if (certificate.getParentId() != 0) {
+			Certificate parent = certificateService.findById(certificate.getParentId());
+			
+			// TODO: 404 if parent is null.
+			
+			req.setParent(parent);
+		}
+		
+		req = certificateRequestService.makeRequest(req);
+		
+		if (certificateRequestService.canAutoAccept(req)) {
+			certificateService.accept(req);
+		}
+		
+		return ResponseEntity.ok(req.toString());
+	}
+	
+	@PutMapping("/request/accept/{id}")
+	public ResponseEntity<String> acceptRequest(@PathVariable Long id) {
+		CertificateRequest req = certificateRequestService.findByIdAndStatusEquals(id, Status.PENDING);
+		if (req == null) {
+			return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
+		}
+		
+		Certificate cert = certificateService.accept(req);
+		return ResponseEntity.ok(cert.toString());
+	}
+	
+	@GetMapping("/request/{id}")
+	public ResponseEntity<List<CertificateRequestDto>> getMyRequests(@PathVariable Long id) {
+		User issuer = userService.findById(id);
+		if (issuer == null) {
+			return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
+		}
+		
+		List<CertificateRequest> requests = certificateRequestService.findByIssuer(issuer);
+		List<CertificateRequestDto> result = requests.stream()
+				.map(r -> new CertificateRequestDto(
+						r.getId(), 
+						r.getIssuer().getId(), 
+						r.getValidTo(), 
+						r.getParent() != null ? r.getParent().getId() : 0, 
+						r.getType(), 
+						r.getStatus())
```

Comment:
Consider using `ModelMapper` for this

Reply:
Example? I still don't understand how `ModelMapper` works.

Reply:
Add `ModelMapper` as a bean to the Configuration, autowire it into the controller, then do

```java
List<CertificateRequestDto> result = requests.stream()
		.map(r -> modelMapper.map(r, CertificateRequestDto.class)).toList();
```
It should even be smart enough to correctly map the issuer and parent id. Although I'm not sure what it will do if `getParent()` is null. Worst case scenario you could add a typemap that uses the existing ternary for mapping.

Speaking of, since I forgot to mention it earlier: Consider assigning `null` to `CertificateRequestDto`'s `parentId` instead of 0. Between 0 and null, when talking about (non-primitive) objects, null more clearly represents a lack of value. Although if we wanted to be super explicit, maybe we could somehow make it work with `Optional`?

Reply:
When `getParent()` is null, it seems to just return `null`, which resolves the assignment comment. I'm curious as how it knows to map by id and not some other field. More specifically, I'm worried that when we add a new property to one of our entities, it may cause the mapper to map incorrectly which warrants the need for tests. [d4eb49d](https://github.com/magley/sssscs/pull/3/commits/d4eb49df6f8d640946eccd94b08b62f12e6842b1)

Reply:
> I'm curious as how it knows to map by id and not some other field.

It maps based on field names. You can read about it at https://modelmapper.org/getting-started/